src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Waiting/VisualStudioWaitContext.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+using Microsoft.VisualStudio.ProjectSystem.Waiting;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Waiting
+{
+    internal class VisualStudioWaitContext : IWaitContext
+    {
+        private const int DelayToShowDialogSecs = 2;
+
+        private readonly string _title;
+        private string _message;
+        private bool _allowCancel;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly IVsThreadedWaitDialog3 _dialog;
+
+        public VisualStudioWaitContext(IVsThreadedWaitDialogFactory waitDialogFactory,
```

Comment:
Pass through `IVsUIService<IVsThreadedWaitDialogFactory>` here so that we enforce correct UI thread usage here.